
<h1 align="center">rNMP Enrichment Factor and Frequency calculation with rNMP EF heatmaps</h1>
To obtain rNMP EF in a defined range of genomic regions while giving heatmaps for a specified subtype of genomic regions (e.g. CpG related regions), the set of scripts alaso provide average and median rNMP EF based on cell type information provided for other type of analysis
<!-- Improved compatibility of back to top link: See: https://github.com/othneildrew/Best-README-Template/pull/73 -->
<a name="readme-top"></a>
<!--
-->

[![Commits][Commits-shield]][Commits-url]
[![Contributors][contributors-shield]][contributors-url]
[![Forks][forks-shield]][forks-url]
[![Stargazers][stars-shield]][stars-url]
[![Website][website-shield]][website-url]
[![Issues][issues-shield]][issues-url]
[![MIT License][license-shield]][license-url]
[![LinkedIn][linkedin-shield]][linkedin-url]

<!-- TABLE OF CONTENTS -->
<details>
  <summary>Table of Contents</summary>
  <ol>
    <li><a href="##Installation">Installation</a></li>
      <ul>
        <li><a href="###Getting-the-code">Getting the code</a></li>
        <li><a href="###Creating-the-enviroment-with-required-dependencies">Creating the enviroment with required dependencies</a></li>
        <li><a href="###Additional-Dependencies">Additional Dependencies</a></li>
      </ul>
    </li>
    <li><a href="##Usage">Usage</a></li>
      <ul>
        <li><a href="###Defining-variables">Defining variables</a></li>
        <li><a href="###Preprocessing">Preprocessing</a></li>
        <li><a href="###Loading-Functions">Loading Functions</a></li>
       <li><a href="###Analyzing-and-processing-data-1">Analyzing, processing and visualizing data for TSS/TTS (ends) of ranges provided</a></li>
        <li><a href="###Analyzing-and-processing-data-2">Analyzing, processing and visualizing data for center of ranges provided</a></li>
      </ul>
    <li><a href="##Contributing">Contributing</a></li>
    <li><a href="##License">License</a></li>
    <li><a href="##Contact">Contact</a></li>
    <li><a href="##Citations">Citations</a></li>
  </ol>
</details>

<!-- Installation -->
## Installation
### Getting the code
The development version from [GitHub](https://github.com/) with:
```sh
git clone https://github.com/DKundnani/hu_nucl_ribome_analysis
conda env create -f hu_nucl_ribome_analysis/rNMP_EF/r-env.yml
conda activate r_env
```
### Required dependencies 
* bedtools
* R-ggplot2

### Additional Dependencies
* Input files (bed) containing single nucleotide locations, mainly for rNMP data. (another single nucleotide data can also be experimented on!)
* Reference genome files (.fa and .fai) of the organism being used(Also used to generate bed files)
* Range files in preferably bed format See [https://github.com/DKundnani/Omics-pipelines](https://github.com/DKundnani/Omics-pipelines)
* Library information, see "order" in this repository

<!-- USAGE -->
## Usage
### Defining variables
```bash
range='range.bed' #File that has genomic ranges of interest. Multiple subtypes to be analyzed as a part of same heatmap to be included in the same range file, making sure no overlapping regions exist.
outfolder='path/to/outfolder' #Folder to create output files, if already exists, will be overwritten
subtypecol=7 #column in range.bed that specifies the subtypes. (eg. for CpG related regions, it should specific either islands, shores, shelves,inter-cpg regions )
scripts='path/to/rNMP_EF/' #location where these scripts are downloaded or cloned
libmeta='/path/to/order' #Please see order file format in this repository
bedfiles='/storage/home/hcoda1/5/dkundnani3/p-fstorici3-0/rich_project_bio-storici/Hu_analysis/subnfiltbed/nucl/noXY/*.bed' # List of bed files
genome='/storage/home/hcoda1/5/dkundnani3/p-fstorici3-0/rich_project_bio-storici/reference/hg38/filtered_hg38-nucleus-noXY.fa.fai' #Genome index file

```

### Get rNMP counts using annotate.sh
```bash
mkdir -p $outfolder
bash $scripts/annotate.sh -r ${range} -c -o $outfolder -b $bedfiles
```
[-r file with reference annotations or ranges] 

[-c calculate counts in the referenced annotations] 

[-f calculate fraction of coverage in the reference annotations]

[-s counts are separated based on strand, should be in the 6th column of ranges.bed]

[-o output file name, will be stored in the annotations folder created in the current directory]

[-d annotated files will be separated based on nucleus and mitochondria]

[-b bed file name, wild card enabled for multiple bed files] 


### rNMP EF calculation and heatmap matrix generation
```bash
Rscript $scripts/Subtype_ratio.R -a $outfolder/annotated_counts.tsv -c $outfolder/all_counts.tsv -g $genome -f $libmeta -t $subtypecol -o ${outfolder}/$(basename $range)
```
[-a <annotated file generated by bedtools annotate>]

[-c <counts file indicating total counts in bedfiles used for rNMP Ef normalization>]

[-g <genome.fai file indicating chromosome length, the chromosome list should match that in bed files>]

[-s <used if same or opposite strand counts being used, important to calculate correct enrichment factor for single strand data>]

[-f <order in which user prefers to visualize results. 1stcol:order number; 2ndcol:Filename; 3rdcol:Libraryname 4thcol:Cellline,Samplename, 5thcol:Replicate info>]

[-t <column in the annotated file which is to be used for subtypes [default 6]>]

[-o <output file name [default out]>]

This step outputs various files as following:
* $(basename $range)_regions_EF.tsv - rNMP EF for each library, for each region
* $(basename $range)_regions_EF_avg.tsv -  mean rNMP EF for each celltype in order files, for each region
* $(basename $range)_regions_EF_median.tsv -  median rNMP EF for each celltype in order files, for each region
* $(basename $range)_subtype_EF_avg.tsv -  mean rNMP EF for each subtype, for each library
* $(basename $range)_subtype_freq_avg.tsv -  mean rNMP frequency for each subtype, for each library

### Heatmap Visualization
```bash

Rscript $scripts/Subtype_ration_plots.R -o $order -f $outfolder/$(basename $range)_subtype_EF_avg.tsv -y 3

```
[-o <order in which user prefers to visualize results. 1stcol:order number; 2ndcol:Filename; 3rdcol:Libraryname 4thcol:Cellline,Samplename, 5thcol:Replicate info>]
[-f <subtype_EF_avg.tsv generated in the previous step>]
[-y <maximum range for rNMP EF color scheme in the heatmap [default 6] >]
[-m <minimum range for rNMP EF color scheme in the heatmap [default 0] >]


<!-- CONTRIBUTING -->
## Contributing

Contributions are what make the open source community such an amazing place to learn, inspire, and create. Any contributions you make are **greatly appreciated**.

If you have a suggestion that would make this better, please fork the repo and create a pull request. You can also simply open an issue with the tag "enhancement".
Don't forget to give the project a star! Thanks again!

1. Fork the Project
2. Create your Feature Branch (`git checkout -b feature/AmazingFeature`)
3. Commit your Changes (`git commit -m 'Add some AmazingFeature'`)
4. Push to the Branch (`git push origin feature/AmazingFeature`)
5. Open a Pull Request

<p align="right">(<a href="#readme-top">back to top</a>)</p>


<!-- LICENSE -->
## License

Distributed under the MIT License. See `LICENSE.txt` for more information.

<p align="right">(<a href="#readme-top">back to top</a>)</p>



<!-- CONTACT -->
## Contact
Deepali L. Kundnani - [deepali.kundnani@gmail.com](mailto::deepali.kundnani@gmail.com)    [![LinkedIn][linkedin-shield]][linkedin-url] 
<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- ACKNOWLEDGMENTS -->
## Citations
Use this space to list resources you find helpful and would like to give credit to. I've included a few of my favorites to kick things off!+
* <b> Human ribomes reveal DNA-embedded ribonucleotides as a new type of epigenetic mark. </b>
Deepali Lalchand Kundnani, Taehwan Yang, Tejasvi Channagiri, Penghao Xu, Yeunsoo Lee, Mo Sun, Francisco Martinez-Figueroa, Supreet Randhawa, Ashlesha Gogate, Youngkyu Jeon, Stefania Marsili, Gary Newnam, Yilin Lu, Vivian Park, Sijia Tao, Justin Ling, Raymond Schinazi, Zachary Pursell, Abdulmelik Mohammed, Patricia Opresko, Bret Freudenthal, Baek Kim, Soojin Yi, Nata≈°a Jonoska, Francesca Storici, <i>  bioRxiv </i> 2025, [https://doi.org/10.1101/2025.06.27.661996](https://doi.org/10.1101/2025.06.27.661996)

<p align="right">(<a href="#readme-top">back to top</a>)</p>



<!-- MARKDOWN LINKS & IMAGES -->
<!-- https://www.markdownguide.org/basic-syntax/#reference-style-links -->
[contributors-shield]: https://img.shields.io/github/contributors/DKundnani/hu_nucl_ribome_analysis?style=for-the-badge
[contributors-url]: https://github.com/DKundnani/hu_nucl_ribome_analysis/graphs/contributors
[forks-shield]: https://img.shields.io/github/forks/DKundnani/hu_nucl_ribome_analysis?style=for-the-badge
[forks-url]: https://github.com/DKundnani/hu_nucl_ribome_analysis/forks
[stars-shield]: https://img.shields.io/github/stars/DKundnani/hu_nucl_ribome_analysis?style=for-the-badge
[stars-url]: https://github.com/DKundnani/hu_nucl_ribome_analysis/stargazers
[issues-shield]: https://img.shields.io/github/issues/DKundnani/hu_nucl_ribome_analysis?style=for-the-badge
[issues-url]: https://github.com/DKundnani/hu_nucl_ribome_analysis/issues
[license-shield]: https://img.shields.io/github/license/DKundnani/hu_nucl_ribome_analysis?style=for-the-badge
[license-url]: https://github.com/DKundnani/hu_nucl_ribome_analysis/blob/master/LICENSE.txt
[linkedin-shield]: https://img.shields.io/badge/-LinkedIn-black.svg?style=for-the-badge&logo=linkedin&colorB=555
[linkedin-url]: https://linkedin.com/in/deepalik
[product-screenshot]: images/screenshot.png
[commits-url]: https://github.com/DKundnani/hu_nucl_ribome_analysis/pulse
[commits-shield]: https://img.shields.io/github/commit-activity/t/DKundnani/hu_nucl_ribome_analysis?style=for-the-badge
[website-shield]: https://img.shields.io/website?url=http%3A%2F%2Fdkundnani.bio%2F&style=for-the-badge
[website-url]:http://dkundnani.bio/ 
